### PR TITLE
Run config-changed hook on application config update.

### PR DESCRIPTION
**Pull request by ExternalReality**
## Description of change

Run config changed hook on "application configuration" update.

Why is this change needed?

Config changed hook would run on charm config changes but not application config changes.

## QA steps

Deploy an Ubuntu charm, trust the application, it should trigger config changed.

### CODE REVIEW

**Comment by jameinel** [comment]
I think what we actually need is to break out a new Watcher. We can discuss on a hangout.

File: api/uniter/unit.go [hunk 1]
```diff
@@ -561,11 +561,23 @@ func (u *Unit) ClearResolved() error {
...
 func (u *Unit) WatchConfigSettings() (watcher.NotifyWatcher, error) {
+	return getSettingsWatcher(u, "WatchConfigSettings")
+}
+
+// WatchTrustConfigSettings will return a watcher to monitor at least the trust
+// application configuration settings. This is in contrast to Charm
+// configuration settings watchers which are created with WatchConfigSettings
+// and do not monitor for application configuration settings such as "trust".
+func (u *Unit) WatchTrustConfigSettings() (watcher.NotifyWatcher, error) {
+	return getSettingsWatcher(u, "WatchApplicationConfigSettings")
```
Comment (jameinel) [inline api/uniter/unit.go]: I would like to see this pushed up into the APIServer code.

File: apiserver/facades/agent/uniter/uniter.go [hunk 2]
```diff
@@ -928,6 +928,30 @@ func (u *UniterAPI) ClosePorts(args params.EntitiesPortRanges) (params.ErrorResu
...
+	}
+	result, err := u.WatchSettings(args, watcherFn)
+	if err != nil {
+		return params.NotifyWatchResults{}, errors.Trace(err)
+	}
+
+	return result, nil
+}
+
+func (u *UniterAPI) WatchApplicationConfigSettings(args params.Entities) (params.NotifyWatchResults, error) {
```
Comment (jameinel) [inline apiserver/facades/agent/uniter/uniter.go]: so this name would change.

File: apiserver/facades/agent/uniter/uniter.go [hunk 3]
```diff
@@ -928,6 +928,30 @@ func (u *UniterAPI) ClosePorts(args params.EntitiesPortRanges) (params.ErrorResu
...
+	watcherFn := func(u *state.Unit) (state.NotifyWatcher, error) {
+		return u.WatchApplicationConfigSettings()
```
Comment (jameinel) [inline apiserver/facades/agent/uniter/uniter.go]: but this doesn't have to change.

**Comment by ExternalReality** [comment]
$$MERGE$$

**Comment by ExternalReality** [comment]
[Updating trust configuration does not fire config-changed hook](https://trello.com/c/edqwWqfv/144-updating-trust-configuration-does-not-fire-config-changed-hook)

**Pull request merged**